app/Foliage/GitClone.hs

```diff
@@ -0,0 +1,56 @@
+{-# LANGUAGE DeriveAnyClass #-}
+{-# LANGUAGE DerivingVia #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TypeFamilies #-}
+
+-- | Clone a github repository into a cache directory.
+module Foliage.GitClone (
+  gitClone,
+  addGitCloneRule,
+)
+where
+
+import Development.Shake hiding (doesDirectoryExist)
+import Development.Shake.Classes
+import Development.Shake.FilePath
+import Development.Shake.Rule
+import Foliage.Meta (GitHubRepo, gitHubRepoToString)
+import GHC.Generics (Generic)
+import System.Directory (doesDirectoryExist)
+
+newtype GitClone = GitClone {repo :: GitHubRepo}
+  deriving (Eq, Generic)
+  deriving newtype (NFData)
+
+instance Show GitClone where
+  show GitClone{repo} = "gitClone " <> gitHubRepoToString repo
+
+instance Hashable GitClone
+
+instance Binary GitClone
+
+type instance RuleResult GitClone = FilePath
+
+-- | Clone given repo at given revision into the cache directory and return the working copy path.
+gitClone :: GitHubRepo -> Action FilePath
+gitClone repo = apply1 GitClone{repo}
+
+-- | Set up the 'GitClone' rule with a cache directory.
+addGitCloneRule
+  :: FilePath
+  -- ^ Cache directory
+  -> Rules ()
+addGitCloneRule cacheDir = addBuiltinRule noLint noIdentity run
+ where
+  run :: BuiltinRun GitClone FilePath
+  run GitClone{repo} _old _mode = do
+    let path = cacheDir </> "git" </> gitHubRepoToString repo
+
+    alreadyCloned <- liftIO $ doesDirectoryExist path
```

Comment:
Here you should use that `_old` parameter to store (inside shake itself) information from the rule previous run. From the docs:

> As an example, a typical [BuiltinRun](https://hackage.haskell.org/package/shake-0.19.7/docs/Development-Shake-Rule.html#t:BuiltinRun) will look like:
> 
> ```
> run key oldStore mode = do
>     ...
>     pure $ RunResult change newStore newValue
> ```
> 
> Where you have:
> 
> - key, how to identify individual artifacts, e.g. with file names.
> - oldStore, the value stored in the database previously, e.g. the file modification time.
> - mode, either [RunDependenciesSame](https://hackage.haskell.org/package/shake-0.19.7/docs/Development-Shake-Rule.html#v:RunDependenciesSame) (none of your dependencies changed, you can probably not rebuild) or [RunDependenciesChanged](https://hackage.haskell.org/package/shake-0.19.7/docs/Development-Shake-Rule.html#v:RunDependenciesChanged) (your dependencies changed, probably rebuild).
> - change, usually one of either [ChangedNothing](https://hackage.haskell.org/package/shake-0.19.7/docs/Development-Shake-Rule.html#v:ChangedNothing) (no work was required) or [ChangedRecomputeDiff](https://hackage.haskell.org/package/shake-0.19.7/docs/Development-Shake-Rule.html#v:ChangedRecomputeDiff) (I reran the rule and it should be considered different).
> - newStore, the new value to store in the database, which will be passed in next time as oldStore.
> - newValue, the result that [apply](https://hackage.haskell.org/package/shake-0.19.7/docs/Development-Shake-Rule.html#v:apply) will return when asked for the given key.

The argument `oldStore` is `Maybe ByteString` and `newStore` has to be `ByteString`, `oldStore` is `Nothing` if it is the first time the rule runs.

E.g. you could use that 1) to detect whether you checkout out the repo before 2) perhaps remember what commit you had checkedout (not sure if it's worth it but you can do that).

Reply:
Yeah I thought about using the last fetched revision in the store. It felt like an optimization though and not sure if needed. You think we should do that?